syntax = "proto3";

service Doughnut {
   rpc MakeImmutableBlock(MakeImmutableBlockRequest) returns (Block) {}
   rpc MakeMutableBlock(MakeMutableBlockRequest) returns (Block) {}
   rpc MakeNamedBlock(MakeNamedBlockRequest) returns (Block) {}
   rpc NamedBlockAddress(NamedBlockAddressRequest) returns (NamedBlockAddressResponse) {}
   rpc Fetch(FetchRequest) returns (FetchResponse) {}
   rpc Insert(InsertRequest) returns (InsertResponse) {}
   rpc Update(UpdateRequest) returns (UpdateResponse) {}
   rpc Remove(RemoveRequest) returns (RemoveResponse) {}
}

message MakeMutableBlockRequest {
}

message MakeNamedBlockRequest {
  bytes key = 1;
}

message NamedBlockAddressRequest {
  bytes key = 1;
}

message InsertRequest {
  Block block = 1;
}

message UpdateRequest {
  Block block = 1;
  bool decrypt_data = 2;
}

message MakeImmutableBlockRequest {
  bytes data = 1;
  bytes owner = 2;
}

message FetchRequest {
  bytes address = 1;
  bool decrypt_data = 2;
}

message NamedBlockAddressResponse {
  bytes address = 1;
}

message RemoveRequest {
  bytes address = 1;
}

message Version {
  int64 major = 1;
  int64 minor = 2;
  int64 subminor = 3;
}

message ACLEntry {
  Key key_koh = 1;
  bool read = 2;
  bool write = 3;
  bytes token = 4;
}

message PublicKey
{
  bytes rsa = 1;
}

message Key {
  string type = 1;
  PublicKey public_key = 2;
}

message Block {
  string type = 1;
  bytes address = 2;
  oneof payload {
    bytes data = 3;
    bytes data_plain = 22;
  }
  bytes salt = 4;
  // CHB
  bytes owner = 5;
  // NB,OKB,ACB
  bytes signature = 6;
  // NB
  bytes owner_rsa = 7;
  bytes name = 8;
  // OKB, ACB
  Key key_koh = 9;
  int64 version = 10;
  int64 next_seal_version = 23;
  // ACB
  int64 editor = 11;
  bytes owner_token = 12;
  repeated ACLEntry acl = 13;
  int64 data_version = 14;
  bytes data_signature = 15;
  bool world_readable = 16;
  bool world_writable = 17;
  repeated ACLEntry group_acl = 18;
  repeated int64 group_version = 19;
  bool deleted = 20;
  Version seal_version = 21;
}


message FetchResponse {
  string type = 1;
  Block block = 2;
}

message InsertResponse {
}

message UpdateResponse {
  string type = 1;
  bytes message = 2;
  Block current = 3;
}

message RemoveResponse {
}